### PR TITLE
Java based Presets & Player Bot Fight Styles

### PR DESCRIPTION
**Pull request by tobywisener**
This is work i've been doing in the evenings and weekends.

-New PlayerBotFightStyle system for Presets
-New ItemInSlot entity to define an Item ID and inventory slot index, or Inventory instance to look for the slot index
-New WeaponSwitch class which takes ItemInSlot or CombatSpell, and implements beforeSwitch(), shouldSwitch() and afterSwitch()
-Bot presets can have separate private presets from Players
-Elvemage preset (private) & Player Bot
-Converted all JSON presets to Java instances so they are easier to build, read and extend, cleaned up logic and tasks related to reading them from JSON
-New players get a random preset
-Player bots no longer get assigned other Player bots as targets (BountyHunter)

### CODE REVIEW

File: ElvargServer/src/main/java/com/elvarg/game/GameConstants.java [hunk 1]
```diff
@@ -129,9 +128,11 @@ public class GameConstants {
...
 	public static final PlayerBotDefinition[] PLAYER_BOTS = new PlayerBotDefinition[] {
 			new PlayerBotDefinition("Bot Hello123", new Location(3085, 3528), 0),
-			new PlayerBotDefinition("Bot 1337Pk", new Location(3087, 3530), 1),
-			new PlayerBotDefinition("Bot Kids Ranqe", new Location(3089, 3530), 2),
-			new PlayerBotDefinition("Bot Josh", new Location(3091, 3533), 3),
+			new PlayerBotDefinition("Elvemage", new Location(3093, 3529), 1),
+			new PlayerBotDefinition("Bot 1337Pk", new Location(3087, 3530), 2),
+			new PlayerBotDefinition("Bot Kids Ranqe", new Location(3089, 3530), 3),
+			new PlayerBotDefinition("Bot Josh", new Location(3091, 3533), 4),
+			new PlayerBotDefinition("Bot Josh", new Location(3091, 3533), 5),
```
Comment (tobywisener) [inline ElvargServer/src/main/java/com/elvarg/game/GameConstants.java:135]: Need to change to another name/position Bot Josh is declared twice

File: ElvargServer/src/main/java/com/elvarg/game/entity/impl/playerbot/fightstyle/PlayerBotFightStyle.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import com.elvarg.util.ItemIdentifiers;
+
+public abstract class PlayerBotFightStyle {
+
+    // The weapon the PlayerBot switches back to once none of the weapon switches apply
+    public abstract int getMainWeaponId();
+
+    public abstract WeaponSwitch[] getWeaponSwitches();
+
+    // Determines whether a PlayerBot should eat or not on each of its turns
```
Comment (tobywisener) [inline ElvargServer/src/main/java/com/elvarg/game/entity/impl/playerbot/fightstyle/PlayerBotFightStyle.java:18]: Need to make naming consistent / remove currently unused methods

File: ElvargServer/src/main/java/com/elvarg/game/GameConstants.java (hunk 1 above)
Comment (reinismu) [reply ElvargServer/src/main/java/com/elvarg/game/GameConstants.java:135]: WHy do you comment it? :D Just change the name :)

File: ElvargServer/src/main/java/com/elvarg/game/entity/impl/playerbot/fightstyle/PlayerBotFightStyle.java (hunk 2 above)
Comment (reinismu) [reply ElvargServer/src/main/java/com/elvarg/game/entity/impl/playerbot/fightstyle/PlayerBotFightStyle.java:18]: I would do it in this PR

File: ElvargServer/src/main/java/com/elvarg/game/model/areas/Area.java [hunk 3]
```diff
@@ -1,23 +1,61 @@
...
+
+        if (character.isNpc()) {
+            this.npcs.add(character.getAsNpc());
+        }
+    }
 
-    public abstract void leave(Mobile character, boolean logout);
+    public void leave(Mobile character, boolean logout) {
+        if (character.isPlayerBot()) {
+            this.playerBots.remove(character.getAsPlayerBot());
```
Comment (reinismu) [inline ElvargServer/src/main/java/com/elvarg/game/model/areas/Area.java:46]: This is not efficient. O(n) complexity. I suggest using `HashMap<Integer, PlayerBot>` where key is mobile index

File: ElvargServer/src/main/java/com/elvarg/game/model/areas/impl/BarrowsArea.java [hunk 4]
```diff
@@ -17,6 +17,7 @@ public BarrowsArea() {
 
     @Override
     public void enter(Mobile character) {
+        super.enter(character);
```
Comment (reinismu) [inline ElvargServer/src/main/java/com/elvarg/game/model/areas/impl/BarrowsArea.java:20]: Now we have to always remember to add it. I would rather have `public abstract void postEnter()` and `public final void enter(...) { add npcs and stuff... this.postEnter()}`
Then override `postEnter`

File: ElvargServer/src/main/java/com/elvarg/game/GameConstants.java (hunk 1 above)
Comment (tobywisener) [reply ElvargServer/src/main/java/com/elvarg/game/GameConstants.java:135]: To remind myself later on :D

File: ElvargServer/src/main/java/com/elvarg/game/content/presets/impl/Att60Zerker_94.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
+package com.elvarg.game.content.presets.impl;
+
+import com.elvarg.game.content.presets.Presetable;
+import com.elvarg.game.entity.impl.playerbot.fightstyle.PlayerBotFightStyle;
+import com.elvarg.game.model.Item;
+import com.elvarg.game.model.MagicSpellbook;
+
+import static com.elvarg.util.ItemIdentifiers.*;
+
+public class Att60Zerker_94 extends Presetable {
```
[the remark this answers is not in the document]
Comment (tobywisener) [reply ElvargServer/src/main/java/com/elvarg/game/content/presets/impl/Att60Zerker_94.java:10]: I wanted a separate file for each preset though, having them all in one file would get messy very quickly, it comes down to code readability and maintainability for me. Will try to do something with one for each file

File: ElvargServer/src/main/java/com/elvarg/util/chainedwork/Chain.java [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+/**
+ * A utlity class to chain code functions together which depend on delays like OSS's {@code delay(1)}.
+ * The internal clock/delay system is built upon {@link TaskManager}
+ *
+ * <br>Has similiar fields to {@link Action} but doesnt have Mob as fixed parent generic type
+ *
+ * @author Jak | Shadowrs
+ * @version 25/4/2020
+ */
+public class Chain<T> {
```
[the remark this answers is not in the document]
Comment (tobywisener) [reply ElvargServer/src/main/java/com/elvarg/util/chainedwork/Chain.java:24]: Yeah this was the file I meant to stash, will remove

File: ElvargServer/src/main/java/com/elvarg/game/content/presets/Presetable.java [hunk 7]
```diff
@@ -124,4 +130,6 @@ public boolean isGlobal() {
 	public int getIndex() {
 		return index;
 	}
+
+	public PlayerBotFightStyle getPlayerBotFightStyle() { return null; }
```
[the remark this answers is not in the document]
Comment (tobywisener) [reply ElvargServer/src/main/java/com/elvarg/game/content/presets/Presetable.java:134]: Done - created a HashMap for this in PlayerBot

File: ElvargServer/src/main/java/com/elvarg/game/content/presets/impl/Att60Zerker_94.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tobywisener) [reply ElvargServer/src/main/java/com/elvarg/game/content/presets/impl/Att60Zerker_94.java:10]: Okay I kept them one in each file but defined them once as static variable.

**Pull request closed without merging**